In [1]:
# Imports
import pandas as pd
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn import tree
from statistics import mean

In [2]:
# Load the heart disease dataset:
dataframe = pd.read_csv("../data/heart.csv")
dataframe.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

**Question 1: Data Exploration Questions**
* Question 1.1: How many positive examples of heart disease are in this dataset?
* Question 1.2: How many negative examples (no heart disease) are in this dataset?
* Question 1.3: What is the average age for subjects in this dataset?
* Question 1.4: Which variable is most positively correlated with heart disease?
* Question 1.5: Which variable is most negatively correlated with heart disease?
* Question 1.6: How much does Age correlate with Max Heart Rate (MaxHR)?

**Question 1: Answers**
* Question 1.1:
* Question 1.2:
* Question 1.3:
* Question 1.4:
* Question 1.5:
* Question 1.6:

**The work used to find the answers are shown in the code below**

In [3]:
# Question 1.1: How many positive examples of heart disease are in this dataset?
# Question 1.2: How many negative examples (no heart disease) are in this dataset?
q1_1, q1_2 = 0, 0
for i in dataframe["HeartDisease"]:
    if i == 1:
        q1_1 += 1
    elif i == 0:
        q1_2 += 1
print("Question 1.1: The number of positive examples of heart disease", q1_1)
print("Question 1.2: The number of negative examples of heart disease", q1_2)

# Question 1.3: What is the average age for subjects in this dataset?
print("Question 1.3: The average age for subjects in the dataset i", mean(dataframe["Age"]))

Question 1.1: The number of positive examples of heart disease 508
Question 1.2: The number of negative examples of heart disease 410
Question 1.3: The average age for subjects in the dataset i 53.510893246187365


In [4]:
dataframe.corr()

Age  RestingBP  Cholesterol  FastingBS     MaxHR   Oldpeak  \
Age           1.000000   0.254399    -0.095282   0.198039 -0.382045  0.258612   
RestingBP     0.254399   1.000000     0.100893   0.070193 -0.112135  0.164803   
Cholesterol  -0.095282   0.100893     1.000000  -0.260974  0.235792  0.050148   
FastingBS     0.198039   0.070193    -0.260974   1.000000 -0.131438  0.052698   
MaxHR        -0.382045  -0.112135     0.235792  -0.131438  1.000000 -0.160691   
Oldpeak       0.258612   0.164803     0.050148   0.052698 -0.160691  1.000000   
HeartDisease  0.282039   0.107589    -0.232741   0.267291 -0.400421  0.403951   

              HeartDisease  
Age               0.282039  
RestingBP         0.107589  
Cholesterol      -0.232741  
FastingBS         0.267291  
MaxHR            -0.400421  
Oldpeak           0.403951  
HeartDisease      1.000000

In [5]:
# Question 1.4: Which variable is most positively correlated with heart disease?
print("Question 1.4: Oldpeak is most positively correlated with heart disease.")
# This answer is derived from the table above when running dataframe.corr().
# We know Oldpeak is the variable most positively correlated with heart disease
# because Oldpeak has 0.403951 in the HeartDisease column while the other
# variables are either negative or a lower positive number.

# Question 1.5: Which variable is most negatively correlated with heart disease?
print("Question 1.4: MaxHR is most negatively correlated with heart disease.")
# This answer is derived from the table above when running dataframe.corr().
# We know MaxHR is most negatively correlated with heart disease
# because MaxHR has -0.400421 in the HeartDisease column while the other
# variables are either positive or not as negative.

# Question 1.6: How much does Age correlate with Max Heart Rate (MaxHR)?
print("Question 1.6: Age correlates with MaxHR for -0.400421")
# This answer is derived from the dataframe.corr() table above

Question 1.4: Oldpeak is most positively correlated with heart disease.
Question 1.4: MaxHR is most negatively correlated with heart disease.
Question 1.6: Age correlates with MaxHR for -0.400421


In [6]:
# Choose one X variables to predict from. We can add more variables into X here.
X = dataframe[['Age']]

# y variable is the target label we want to predict.
y = dataframe['HeartDisease']

# Split data into two sections. 70% will be for training, 30% will be for testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42)

In [7]:
def predict_heart_disease_from_age(age: int) -> int:
    return int(age > 100)
predicted_y = [predict_heart_disease_from_age(a) for a in X['Age']]
accuracy = accuracy_score(predicted_y, y)
print(f"The predict_heart_disease_from_age accurately classifies {accuracy*100:.2f}% of examples")

The predict_heart_disease_from_age accurately classifies 44.66% of examples


**Question 2: Manual Classifier**
How would we manually create a function to solve this task for us? Try it.
* Question 2.1: The function predict_heart_disease_from_age only predicts indivduals above 100 years old will have heart disease. This won't match any examples in the dataset, but still achieves 44.66% accuracy. Why is that?
* Question 2.2: Try to make the best predict_heart_disease_from_age function you can. What is the best value to split at? How do you pick this? Please include the train and test accuracy.